### ISSUE TITLE
Option B: Fine-tuned MTG Model (LoRA) for EDH Assistant

### ISSUE THREAD

**Issue by mnesler** [mnesler/attractor-tcg-solid-js#3]
# Option B: Fine-tuned MTG Model with LoRA for EDH Deck Building Assistant

## Overview
Use a fine-tuned Magic: The Gathering-specific model (with LoRA adapters) that has deep MTG knowledge baked into its weights. Requires model hosting but provides specialized MTG understanding.

## Architecture
```
User Query → Fine-tuned MTG Model (Mistral-7B + LoRA) → MTG-Specialized Response
```

### Option B1: TrevorJS/mtg-mistral-7b-instruct-sft-merged
- **Size**: 7B parameters
- **Base**: Mistral-7B-Instruct
- **Training**: Supervised Fine-Tuning (SFT) on MTG data
- **Usage**: 8+ Hugging Face Spaces
- **Pros**: Larger model, better reasoning, instruction-following
- **Cons**: Requires GPU hosting (16GB+ VRAM)

### Option B2: minimaxir/magic-the-gathering-flan-t5-xl
- **Size**: ~3B parameters (Flan-T5-XL)
- **Base**: Flan-T5-XL (encoder-decoder)
- **Training**: LoRA fine-tuning (~9M adapter params)
- **Dataset**: Cards through "Phyrexia: All Will Be One" (Feb 2023)
- **Features**: Hierarchical sampling, subset sampling
- **Pros**: Efficient LoRA approach, good at structured generation
- **Cons**: Focused on card generation, not strategy advice

### Option B3: minimaxir/magic-the-gathering (GPT-2)
- **Size**: ~1M parameters (tiny GPT-2)
- **Training**: 8 hours on V100, ~22k encoded cards
- **Pros**: Extremely lightweight, fast inference
- **Cons**: Too small for complex reasoning, only card generation

### Pipeline Nodes

1. **load_model** (setup - outside pipeline)
   - Load `TrevorJS/mtg-mistral-7b-instruct-sft-merged`
   - Options for hosting:
     - Local GPU server (vLLM, TGI)
     - RunPod/Replicate
     - Hugging Face Inference Endpoints
   - Set up OpenAI-compatible endpoint

2. **parse_deck_context** (codergen)
   - Extract commander and strategy
   - Parse current decklist
   - Identify color identity

3. **query_mtg_model** (codergen)
   - Use fine-tuned Mistral model
   - Endpoint: `http://localhost:8000/v1/chat/completions` (local vLLM)
   - System prompt: EDH deck building consultant
   - Temperature: 0.7
   - Max tokens: 2048
   - Model understands MTG terminology, card names, mechanics

4. **validate_suggestions** (codergen - optional)
   - Cross-check suggested cards exist
   - Verify Commander legality
   - Check color identity matches

5. **format_response** (codergen)
   - Structure advice with reasoning
   - Add Scryfall links for suggested cards

## System Prompt for Fine-tuned Model

```
You are an EDH/Commander deck building expert. The user has the following deck:

{{deck_context}}

User question: {{user_message}}

Provide specific card suggestions with reasoning. Focus on:
- Synergies with the commander
- Mana curve considerations
- Card advantage and interaction
- Win conditions

Be concise and actionable.
```

### Local Hosting (Recommended for Development)

**vLLM Setup:**
```bash
# Install vLLM
pip install vllm

# Run server
python -m vllm.entrypoints.openai.api_server \
  --model TrevorJS/mtg-mistral-7b-instruct-sft-merged \
  --host 0.0.0.0 \
  --port 8000 \
  --dtype float16

# Test
curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "TrevorJS/mtg-mistral-7b-instruct-sft-merged",
    "messages": [{"role": "user", "content": "Best cards for Atraxa?"}],
    "max_tokens": 512
  }'
```

**Requirements:**
- GPU: NVIDIA with 16GB+ VRAM (RTX 4090, A100, etc.)
- RAM: 32GB+ system RAM
- Storage: 15GB for model weights

### Cloud Hosting Options

1. **RunPod** (~$0.40/hour for RTX 4090)
   - Deploy vLLM container
   - Pay per hour of usage
   - Good for testing/low traffic

2. **Replicate** (pay-per-inference)
   - ~$0.001 per inference
   - No idle costs
   - Good for production

3. **Hugging Face Inference Endpoints**
   - ~$0.60/hour for dedicated GPU
   - Managed infrastructure
   - Auto-scaling available

## Implementation Steps

1. **Set up model hosting** (choose local/cloud)
2. Create `pipelines/edh-assistant-option-b.dot`
3. Update `src/lib/server/chat.ts`
4. Configure model endpoint in `.env`:
   ```
   MTG_MODEL_ENDPOINT=http://localhost:8000/v1/chat/completions
   MTG_MODEL_NAME=TrevorJS/mtg-mistral-7b-instruct-sft-merged
   ```
5. Test with MTG-specific queries

## Data Flow Example

```yaml
Input:
  deck_context: "Commander: Atraxa, Praetors' Voice. Cards: Sol Ring, Command Tower..."
  user_message: "What are some good proliferate cards?"

Step 1 - Parse Context:
  commander: "Atraxa, Praetors' Voice"
  colors: [W,U,B,G]
  strategy: "Proliferate"

Step 2 - Query MTG Model:
  POST http://localhost:8000/v1/chat/completions
  Response: "For Atraxa proliferate, consider:
  - Flux Channeler: Triggers on noncreature spells
  - Karn's Bastion: Utility land slot
  - Contagion Engine: Board control + proliferate
  - Inexorable Tide: Passive proliferate on spell cast
  ..."

Step 3 - Validate (Optional):
  Check cards exist in Scryfall
  Verify legality

Step 4 - Format:
  Add Scryfall links, mana costs, brief descriptions
```

## Training Your Own LoRA (Optional)

If existing models aren't sufficient:

### Dataset Creation
```python
# Collect training data
from datasets import load_dataset

# Option 1: Use existing MTG dataset
mtg_cards = load_dataset("MechaCroc/magic-the-gathering")

# Option 2: Scrape EDH data
# - EDHREC deck lists
# - r/EDH discussions
# - Commander gameplay articles

# Format as instruction dataset
data = [
  {
    "instruction": "Suggest cards for Atraxa superfriends",
    "input": "Commander: Atraxa. Strategy: Planeswalker proliferate",
    "output": "Top suggestions: 1. Deepglow Skate..."
  },
  ...
]
```

### LoRA Fine-tuning
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import LoraConfig, get_peft_model

# Load base model
model = AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-Instruct-v0.3")
tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-Instruct-v0.3")

# LoRA config
lora_config = LoraConfig(
    r=16,  # LoRA rank
    lora_alpha=32,
    target_modules=["q_proj", "v_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM"
)

# Apply LoRA
model = get_peft_model(model, lora_config)

# Fine-tune with your MTG dataset
# (Use HuggingFace Trainer or custom training loop)
```

**Training Resources:**
- GPU: A100 40GB (rent on RunPod/Lambda)
- Time: 2-8 hours depending on dataset size
- Cost: ~$10-40 for full training run

## Pros
✅ Deep MTG knowledge baked into weights
✅ Understands card names, mechanics, interactions
✅ No need for RAG/context enrichment
✅ Works offline (if hosted locally)
✅ Can specialize on EDH-specific strategy

## Cons
❌ Requires GPU hosting (cost/complexity)
❌ Training data cutoff (gets outdated with new sets)
❌ Less general reasoning than large frontier models
❌ Inference slower than API calls
❌ Model may hallucinate fake cards
❌ Needs manual updates for new sets

### Local Hosting
- **Hardware**: $1500-3000 (RTX 4090 or used A100)
- **Power**: ~$50/month (24/7 operation)
- **Maintenance**: Time investment

### Cloud Hosting
- **RunPod**: $0.40/hour × 730 hours = $292/month (24/7)
- **Spot instances**: ~$100-150/month (60% savings)
- **Replicate**: $0.001 × queries/month

### Comparison to API (Kimi k2)
- Kimi k2 via OpenRouter: ~$0.01 per query
- Break-even: 10,000+ queries/month favors self-hosting

## Success Metrics
- Card suggestion accuracy (are cards real and legal?)
- MTG terminology understanding
- Synergy reasoning quality
- Inference latency (<2s)
- Hallucination rate (<5%)

## Related Files
- `src/lib/server/chat.ts` - Chat integration
- `.env` - Model endpoint configuration
- `docker-compose.yml` - Potential vLLM deployment

## Dependencies
- GPU with 16GB+ VRAM (for Mistral-7B)
- vLLM or Text Generation Inference
- Hugging Face Transformers
- Attractor MCP
